lib/receptacle.rb

```diff
@@ -1,14 +1,47 @@
 # frozen_string_literal: true
 
 require "receptacle/version"
-require "receptacle/interface_methods"
-require "receptacle/method_delegation"
+require "receptacle/repository"
 
 module Receptacle
   module Repo
+    module ClassMethods
+      def mediate(method_name)
+        define_singleton_method(method_name) do |*args, **kwargs|
+          raise NotConfigured unless @strategy
+
+          with_wrappers(@wrappers.dup, method_name, *args, **kwargs) do |*sub_args, **sub_kwargs|
+            strategy.new.public_send(method_name, *sub_args, **sub_kwargs)
+          end
+        end
+      end
+
+      def wrappers(wrappers)
+        @wrappers = wrappers
+      end
+
+      def strategy(value = nil)
+        if value
+          @strategy = value
+        else
+          @strategy
+        end
+      end
+
+      def with_wrappers(wrappers, method_name, *args, **kwargs, &block)
+        next_wrapper = wrappers.shift
+        if next_wrapper
+          next_wrapper.new.public_send(method_name, *args, **kwargs) do |*sub_args, **sub_kwargs|
+            with_wrappers(wrappers, method_name, *sub_args, **sub_kwargs, &block)
```

Comment:
how do you do the before/after now ?

Reply:
you would do it like this, e.g. for sending an update message:

```ruby
let(:add5_wrapper) do
  Class.new do
    def update(id:, changeset:)
      before = Repositories::Foo.find(id: id)
      after = yield(id: id, changeset: changeset)
      Messages.publish(build_update_message(before, after)
    end
  end
end
```

Reply:
its basically like a rack middleware iirc

Reply:
in comparison with the old implementation:

```ruby
class MessagesExistingInterface
  def before_update(**args)
    store_before_operation(args)
    { **args }
  end

  def after_update(model, _args)
    changed_attributes = Helper::ChangedAttributes.build(model,
                                                         @before_operation_model)
    return model if changed_attributes.empty?

    MessagePublisher.publish(event:              :updated,
                             entity:             model,
                             changed_attributes: changed_attributes,
                             opts:               opts)
    model
  end

  private

  def store_before_operation(args, model = nil)
    @before_operation_model = model || find_entity(**args) # find_entity has to be defined in service wrapper
  rescue NoMethodError
    nil
  end
end

class MessagesNewInterface
  def update(id:, changeset:)
    before_model = find_entity(id: id)
    yield(id: id, changest: changeset).tap do |model|
      changed_attributes = Helper::ChangedAttributes.build(model, before_model)
      return model if changed_attributes.empty?

      MessagePublisher.publish(event: :updated,
                                                            entity:             model,
                                                            changed_attributes: changed_attributes)
      model
    end
  end
end
```